src/mds/MDCache.cc

```diff
@@ -8441,9 +8439,19 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,
       if (dn->state_test(CDentry::STATE_PURGING))
 	return -CEPHFS_ENOENT;
 
+      CDentry::linkage_t *dnl = dn->get_projected_linkage();
+      // If an auth check was deferred before and the target inode is found
+      // not to exist now, do the auth check here if necessary.
+      if (want_auth && want_dentry && want_inode && depth == path.depth() - 1 &&
+	  dnl->is_null() && (r = maybe_request_forward_to_auth(mdr, cf, dn)) != 0)
+	return r;
+
       if (rdlock_path) {
 	lov.clear();
-	if (xlock_dentry && depth == path.depth() - 1) {
+	// do not xlock the tail dentry if target inode exists and caller wants it
+	if (xlock_dentry && (dnl->is_null() || !want_inode) &&
```

Comment:
doesn't the dentry linkage need to be valid here as well ?

Reply:
> doesn't the dentry linkage need to be valid here as well ?

The tail dentry can be `null` and we can also create one if it doesn't exist. The caller gets a `null` tail dentry and then links it with an inode, a file or hard link can be created. And we need to provide a xlocked auth tail dentry to the caller because it needs to be modified.